helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java

```diff
@@ -520,8 +520,12 @@ private void loadRebalance(Resource resource, Partition partition,
        ResourceControllerDataProvider cache,
       boolean onlyDownwardLoadBalance, StateModelDefinition stateModelDefinition, Set<String> messagesThrottled,
       Map<Partition, List<Message>> resourceMessageMap) {
-    if (onlyDownwardLoadBalance && isLoadBalanceDownwardStateTransition(messageToThrottle, stateModelDefinition)) {
-      // Remove the message already allowed for downward state transitions.
+    // If only downward allowed: 1) any non-downward ST messages will be throttled and removed.
+    //                           2) any downward ST messages will respect the throttling.
+    // If not only downward allowed, all ST messages should respect the throttling.
+    if (onlyDownwardLoadBalance && !isLoadBalanceDownwardStateTransition(messageToThrottle, stateModelDefinition)) {
```

Comment:
The change makes sense now. But ideally, I think we should let the throttling class handle extra logic like "onlyDownwardLoadBalance". I propose that we refactor this when deprecating the IntermediateStateCalcStage class. Please add a TODO here if you think it makes sense.

Reply:
Yeah. That makes sense to me. Will change it.